hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/CellBlocks.java

```diff
@@ -0,0 +1,427 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Cellersion 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY CellIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hbase.regionserver;
+
+import org.apache.hadoop.hbase.Cell;
+
+import java.util.*;
+import java.util.concurrent.ConcurrentNavigableMap;
+
+/**
+ * CellBlocks stores a constant number of elements and is immutable after creation stage.
+ * Due to being immutable the CellBlocks can be implemented as array.
+ * The actual array is on- or off-heap and is implemented in concrete class derived from CellBlocks.
+ * The CellBlocks uses no synchronization primitives, it is assumed to be created by a
+ * single thread and then it can be read-only by multiple threads.
+ */
+public abstract class CellBlocks implements ConcurrentNavigableMap<Cell,Cell> {
+
+  private final Comparator<? super Cell> comparator;
+  private int minCellIdx   = 0;   // the index of the minimal cell (for sub-sets)
+  private int maxCellIdx   = 0;   // the index of the maximal cell (for sub-sets)
+  private boolean isDescending = false; // array can be easily traversed backward in descending order
+
+  /* C-tor */
+  public CellBlocks(Comparator<? super Cell> comparator, int min, int max, boolean d){
+    this.comparator = comparator;
+    this.minCellIdx = min;
+    this.maxCellIdx = max;
+    this.isDescending = d;
+  }
+
+  /* Used for abstract CellBlocks creation, implemented by derived class */
+  protected abstract CellBlocks createCellBlocks(Comparator<? super Cell> comparator, int min,
+      int max, boolean d);
+
+  /* Assuming array underneath implementation this comes instead of array[i] */
```

Comment:
this comment is redundant :) 
no need to have any assumption on the concrete class 
can say instead returns the i-th cell in the cell block